Allow Inspect.Algebra.to_doc/2 Options To Be Overriden

Legrandk opened the pull request:
## PROBLEM

When serializing structs that are somehow opaque, like `Version` or `Money`, we found `Code.eval_string/3` to not be able to evaluate such kinds of serialized values using the current `%Inspect.Opts{}` values passed to the `Inspect.Algebra.to_doc/2` function:

Example:
```elixir
> struct = %{key: Version.parse!("1.2.3")}
%{key: #Version<1.2.3>}
> serialized_value = struct |> Inspect.Algebra.to_doc(%Inspect.Opts{limit: :infinity, printable_limit: :infinity, pretty: true}) |> Inspect.Algebra.group() |> Inspect.Algebra.format(80) |> Enum.join()
"%{key: #Version<1.2.3>}"
> Code.eval_string(serialized_value, [], __ENV__)
** (TokenMissingError) iex:20: missing terminator: } (for "{" starting at line 20)
```

## SOLUTION
We should improve how we pass the `%Inspect.Opts{}` to the `Inspect.Algebra.to_doc/2` so see can customize it by setting the `structs` param to be `false`. Now the structs won't be formatted by the inspect protocol but printed as regular maps. Notice the default value is `true`.

Thanks to this optional parameter, the serialized value will look as:
```elixir
"%{\n  key: %{__struct__: Version, build: nil, major: 1, minor: 2, patch: 3, pre: []}\n}"
```

instead of
```elixir
"%{key: #Version<1.2.3>}"
```

### TESTING

We found that following the `No` branch from the `Snap File Exists?` comparison is good enough to test the new functionality, as we need just to ensure how the struct is serialized at all.
```
                             ┌──────────────────────────────┐
                             │  match(actual_value, caller) │
                             └──────────────┬───────────────┘
                                            │
                                            │
                          NO       ┌────────▼──────────┐    YES
                 ┌─────────────────┤ snapfile_exists?  ├───────────────┐
                 │                 └───────────────────┘               │
                 │                                                     │
                 │                                                     │
   ┌─────────────▼──────────────┐                           ┌──────────▼────────┐
   │ create_empty_snapshot_file │                           │ get_snapshot(file)│
   └─────────────┬──────────────┘                           └──────────┬────────┘
                 │                                                     │
                 │                                                     │
┌────────────────▼──────────────────┐            ┌─────────────────────▼──────────────────────┐
│ save_snapshot(file, actual_value) │            │ assert(file, snapshot_value, actual_value) │
└────────────────┬──────────────────┘            └─────────────────────┬──────────────────────┘
                 │                                                     │
                 │                                                     │
               ┌─▼─┐                                     ┌─────────────▼─────────────────┐  YES      ┌───┐
               │:OK│                                     │ snapshot_value == actual_value├───────────►:OK│
               └───┘                                     └─────────────┬─────────────────┘           └───┘
                                                                       │NO
                                                                       │
                                                                  ┌────▼──────┐   MO                 ┌────────┐
                                                                  │ override? ├──────────────────────► :ERROR │
                                                                  └────┬──────┘                      └────────┘
                                                                       │
                                                                       │YES
                                                                       │
                                                         ┌─────────────▼───────────────────┐
                                                         │save_snapshot(file, actual_value)│
                                                         └─────────────┬───────────────────┘
                                                                       │
                                                                       │
                                                                      ┌▼──┐
                                                                      │:OK│
                                                                      └───┘
```

legrandk-jt committed:
[WORK-289] Inspect.Algebra.to_doc/1 Options Can Be Overriden

DCzajkowski commented:
Hello, [@Legrandk](https://github.com/Legrandk)! Thank you very much for your contribution. I will try to review this and release in the next two days. (:

Legrandk commented:
Hi @DCzajkowski ! Perfect! Let me know if you need something else :)

DCzajkowski commented on test/unit/snapshy_test.exs:
```diff
@@ -24,6 +24,63 @@ defmodule SnapshyTest do
+
+    assert File.rm(snap_filename)
+
+    unless original_serialized_inspect_options do
```
Shouldn't this be an if?

```suggestion
    if original_serialized_inspect_options do
```

However, I think resetting every time to the original without any `if` branch would be more desired. What do you think?

Legrandk replied:
Done: 195ee237e827e4522f75d36bf72f2ccda556d886

DCzajkowski commented on test/unit/snapshy_test.exs:
```diff
@@ -24,6 +24,63 @@ defmodule SnapshyTest do
+
+    assert {:ok, serialized_value} = File.read(snap_filename)
+
+    assert {value, []} = Code.eval_string(serialized_value, [], __ENV__)
```
`value` is unused, so we should prefix it with `_`.

```suggestion
    assert {_value, []} = Code.eval_string(serialized_value, [], __ENV__)
```

However, don't we want to test here `opaque_struct` is equal to `value`? I know checking it serialized without error is good enough, but that would add extra confidence in its serialization effectiveness.

Legrandk replied:
Yes, double-checking the value against the original opaque struct will make it more robust, imo.

DCzajkowski commented on lib/snapshy.ex:
```diff
@@ -195,6 +191,21 @@ defmodule Snapshy do
+      limit: :infinity,
+      printable_limit: :infinity,
+      pretty: true
+    ]
```
Shouldn't we make `structs: false` the default?

I think it will not introduce a regression, because:
1. All existing snapshots will pass,
2. All new snapshots will be generated with the new, more robust method,
3. When using `SNAPSHY_OVERRIDE=true` only failing snapshots will be overridden, so we won't have any unintentional snapshot updates.

Let me know what you think.

Legrandk replied:
Totally agree! :)

I haven't used this at first because I didn't want to change the original configuration used by the library  [`structs: true`](https://hexdocs.pm/elixir/Inspect.Opts.html). 

`1:` Right.
`2:` Right.
`3:` Right.

DCzajkowski reviewed (commented):
Sorry for taking my time to review. Had quite a busy time at work this week.

Let me know what you think :)

legrandk-jt committed:
[WORK-289] Pull Request Suggestions Added

- Default `structs` value now is `false`
- Pattern matching used by the test compares with expected struct.
- Removed branch statement.

DCzajkowski commented:
This looks great! Thank you for your contribution!

DCzajkowski commented:
New version [0.3.0](https://github.com/DCzajkowski/snapshy/releases/tag/0.3.0) released 🎉

Legrandk commented:
Many thanks Dariusz for including this little change :)
